Add `--site-isolation` option to TestWebKitAPI

beidson opened the pull request:
#### bbafb05be33d6f01078d44067866ff8cf8a79b36
<pre>
Add `--site-isolation` option to TestWebKitAPI
<a href="https://rdar.apple.com/149976547">rdar://149976547</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292026">https://bugs.webkit.org/show_bug.cgi?id=292026</a>

Reviewed by Alex Christensen.

All API tests eventually need to work with site isolation on, and exploring which ones
currently fail is useful in identifying areas that need work.

This patch adds a command line option to trivially force all WKWebView&apos;s to enable site
isolation for testing and feature development.

When we get to the point where site isolation can be enabled by default globally, we
can remove (or invert!) this command line option.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(+[WKPreferences _forceSiteIsolationAlwaysOnForTesting]):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::WebPreferences):
(WebKit::WebPreferences::forceSiteIsolationAlwaysOnForTesting):
* Source/WebKit/UIProcess/WebPreferences.h:
* Tools/TestWebKitAPI/mac/mainMac.mm:
(forceSiteIsolationForTesting):
(handleArguments):
(main):
(buildArgumentDefaults): Deleted.

Canonical link: <a href="https://commits.webkit.org/294094@main">https://commits.webkit.org/294094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd909924bcf81c4d0fd71063f2ca2a73487bc8ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100799 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10750 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105936 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51388 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102840 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28925 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/33788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15953 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91046 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/57104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15763 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/9055 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/50763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9129 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108291 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27917 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28280 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87248 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85253 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21699 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29967 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7695 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/21921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27852 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33108 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27663 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30981 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29221 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/mac/mainMac.mm:
```diff
@@ -30,7 +30,18 @@
+{
+#if !defined(BUILDING_TEST_IPC) && !defined(BUILDING_TEST_WTF) && !defined(BUILDING_TEST_WGSL)
+    SEL sel = NSSelectorFromString(@"_forceSiteIsolationAlwaysOnForTesting");
+    if ([[WKPreferences class] respondsToSelector:sel])
```
This shouldn't be necessary.

achristensen07 commented on Tools/TestWebKitAPI/mac/mainMac.mm:
```diff
@@ -47,6 +58,8 @@ void buildArgumentDefaults(int argc, char** argv, NSMutableDictionary *argumentD
             argumentDefaults[@"WebKit2GPUProcessForDOMRendering"] = @YES;
         else if (strcmp(argv[i], "--no-use-gpu-process") == 0)
             argumentDefaults[@"WebKit2GPUProcessForDOMRendering"] = @NO;
+        else if (strcmp(argv[i], "--force-site-isolation") == 0)
```
--site-isolation instead of --force-site-isolation would be slightly preferable to me as it matches the run-webkit-tests argument.

achristensen07 commented on Tools/TestWebKitAPI/mac/mainMac.mm:
```diff
@@ -30,7 +30,18 @@
+#if !defined(BUILDING_TEST_IPC) && !defined(BUILDING_TEST_WTF) && !defined(BUILDING_TEST_WGSL)
+    SEL sel = NSSelectorFromString(@"_forceSiteIsolationAlwaysOnForTesting");
+    if ([[WKPreferences class] respondsToSelector:sel])
+        [[WKPreferences class] performSelector:sel];
```
nit: I like dot syntax for WKPreferences.class

beidson committed:
Add `--site-isolation` option to TestWebKitAPI
rdar://149976547
https://bugs.webkit.org/show_bug.cgi?id=292026

Reviewed by Alex Christensen.

All API tests eventually need to work with site isolation on, and exploring which ones
currently fail is useful in identifying areas that need work.

This patch adds a command line option to trivially force all WKWebView's to enable site
isolation for testing and feature development.

When we get to the point where site isolation can be enabled by default globally, we
can remove (or invert!) this command line option.

* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(+[WKPreferences _forceSiteIsolationAlwaysOnForTesting]):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::WebPreferences):
(WebKit::WebPreferences::forceSiteIsolationAlwaysOnForTesting):
* Source/WebKit/UIProcess/WebPreferences.h:
* Tools/TestWebKitAPI/mac/mainMac.mm:
(forceSiteIsolationForTesting):
(handleArguments):
(main):
(buildArgumentDefaults): Deleted.

Canonical link: https://commits.webkit.org/294094@main
